docs: Update Wave use cases and FAQ (#917)
seqeralabs/wave: docs/nextflow/use-cases.md

---
title: Use cases
description: Learn how to use Wave with Nextflow for container management, building, and security scanning
date: "2024-08-22"
last_update: "2025-10-16"
tags: [nextflow, wave, use cases]
---

> With Nextflow and Wave, you can build, upload, and manage the container images automatically and on-demand during pipeline execution.

Reviewer: ```suggestion
With Nextflow and Wave, you can build, upload, and manage the container images automatically and on demand during pipeline execution.
```

Nextflow integration with Wave requires Nextflow 22.10.0 or later.
:::

## Access private container repositories

> Wave to access private repositories for your Nextflow pipelines.

Reviewer: ```suggestion
Use Wave to access private repositories for your Nextflow pipelines.
```
Else Wave is a verb here.

    ```

1. (Optional) Set your Wave strategy to prioritize Dockerfiles in your Nextflow configuration:

    ```groovy
    wave.strategy = ['dockerfile','container']
    ```

    :::note
    The `container` directive takes precedence over a Dockerfile by default.
    :::

:::warning
Wave does not support `ADD`, `COPY`, or other Dockerfile commands that access files in the host file system.
:::

</details>

## Build Conda-based containers

Wave can provision containers based on the [`conda` directive](https://www.nextflow.io/docs/latest/process.html#conda).
This allows you to use Conda packages in your pipeline, even on cloud-native platforms like AWS Batch and Kubernetes, which do not support the Conda package manager directly.

<details open>
<summary>**Build Conda-based containers**</summary>

To enable Wave to provision Conda package containers:

1. Define your Conda packages using the `conda` directive in your pipeline processes.

1. Enable Wave in your Nextflow configuration:

    ```groovy
    wave.enabled = true
    ```

1. (Optional) Set your Wave strategy to prioritize `conda` in your Nextflow configuration:

    ```groovy
    wave.strategy = ['conda']
    ```

>     :::note

Reviewer: The back-to-back :::note admonitions won't look great. Can you either combine them into one with bullets, or make one of them an :::info ?

Author: I didn't feel good about this either. Making the second point an `info`

    :::

1. Set your Conda channel priority:

    ```groovy
    conda.channels = '<CONDA_CHANNELS>'
    ```

    Replace `<CONDA_CHANNELS>` with a comma-separated list of your channel priorities. For example, `seqera,conda-forge,bioconda,defaults`.

</details>

## Build Singularity containers

Nextflow can build Singularity native images on demand using a `Singularityfile` or Conda packages.
Images are uploaded to an OCI-compliant container registry of your choice and stored as an [ORAS artifact](https://oras.land/).

:::note
Requires Nextflow version 23.09.0-edge or later and a version of Singularity (or Apptainer) that supports pulling images using the `oras:` pseudo-protocol.
:::

<details open>
<summary>**Build Singularity containers**</summary>

To enable provisioning of Singularity images:

1. Add the following to your Nextflow configuration:

    ```groovy
    wave.enabled = true
    wave.freeze = true
    wave.strategy = ['conda']
    singularity.enabled = true
    ```

1. (Optional) To store your Singularity image files in a private registry:

>     1. Configure you repository access in Seqera. See [Seqera Platform credentials](https://docs.seqera.io/platform/latest/credentials/overview) for more information.

Reviewer: ```suggestion
    1. Configure your repository access in Seqera. See [Seqera Platform credentials](https://docs.seqera.io/platform-cloud/credentials/overview) for more information.
```

1. Add your build repository to your Nextflow configuration:

    ```groovy
    wave.build.repository = '<BUILD_REPOSITORY>'
    ```

    Replace `<BUILD_REPOSITORY>` with your container registry.

</details>

## Security scan containers

Wave scans containers used in your Nextflow pipelines for security vulnerabilities. This feature helps you ensure that your pipelines use secure container images by identifying potential security risks before and during pipeline execution.

<details open>
<summary>**Security scan containers**</summary>

To enable container security scanning:

1. Add the following to your Nextflow configuration:

        ```groovy
        wave.enabled = true
        wave.scan.mode = 'required'
        tower.accessToken = '<TOWER_ACCESS_TOKEN>'
        ```

    Replace `<TOWER_ACCESS_TOKEN>` with your [Seqera access token](../tutorials/nextflow-wave.mdx#create-your-seqera-access-token).

1. Add the acceptable vulnerability levels to your Nextflow configuration:

    ```groovy
    wave.scan.allowedLevels = 'low,medium'
    ```

    Accepted vulnerability levels include: `low`, `medium`, `high`, and `critical`.

:::note

> When you set `wave.scan.mode` to `required`, Wave blocks pipeline execution if containers contain vulnerabilities above the specified threshold.

Author: ```suggestion
When you set `wave.scan.mode` to `required`, Wave blocks pipeline execution if containers have vulnerabilities above the specified threshold.
```
